applications/ShapeOptimizationApplication/python_scripts/analyzer_internal.py

```diff
@@ -70,25 +70,28 @@ def AnalyzeDesignAndReportToCommunicator( self, currentDesign, optimizationItera
 
     # --------------------------------------------------------------------------
     def FinalizeAfterOptimizationLoop( self ):
-        for function in self.response_functions.values():
-            function.Finalize()
+        for _, response in self.response_functions:
+            response.Finalize()
 
     # --------------------------------------------------------------------------
     @staticmethod
     def __CreateResponseFunctions( specified_responses, model ):
         available_csm_response_functions = ["strain_energy", "mass", "eigenfrequency"]
 
-        response_functions = {}
+        response_functions = []
+        response_ids = []
 
-        for (response_id, response_settings) in specified_responses:
-            if response_id in response_functions.keys():
+        for response_id, response_settings in specified_responses:
+            if response_id in response_ids:
                 raise NameError("There are multiple response functions with the following identifier: " + response_id)
 
-            if response_settings["response_type"].GetString() in available_csm_response_functions:
-                response_functions[response_id] = csm_response_factory.CreateResponseFunction(response_id, response_settings, model)
-            else:
+            if response_settings["response_type"].GetString() not in available_csm_response_functions:

```

Comment:
the previous version:

```python
if in csm_response_functions:
    # create...
else:
    # error
```

was more flexible to be extendend in the future for e.g. potential flow or fluid reponses.

Reply:
You are right.